pyfar/dsp/dsp.py

```diff
@@ -187,3 +187,109 @@ def spectrogram(signal, dB=True, log_prefix=20, log_reference=1,
     times -= times[0]
 
     return frequencies, times, spectrogram
+
+
+def regularized_spectrum_inversion(
+        signal, freq_range,
+        regu_outside=1., regu_inside=10**(-200/20), regu_final=None):
+    """Invert the spectrum of a signal applying frequency dependent
+    regularization. Regularization can either be specified within a given
+    frequency range using two different regularization factors, or for each
+    frequency individually using the parameter `regu_final`. In the first case
+    the regularziation factors for the frequency regions are cross-faded using
+    a raised cosine window function. Note that in the latter case all
+    remaining options are ignored and an array matching the number of
+    frequency bins of the signal needs to be given.
+
+    Parameters
+    ----------
+    signal : pyfar.Signal
+        The signals which spectra are to be inverted.
+    freq_range : tuple, array_like, double
+        The upper and lower frequency limits outside of which the
+        outside regularization factor is to be applied.
+    regu_outside : float, optional
+        The regularization factor outside the frequency range, by default 1.
+    regu_inside : float, optional
+        The regularization factor inside the frequency range, by
+        default 10**(-200/20)
+    regu_final : float, array_like, optional
+        The final regularization factor for each frequency, by default None.
+        If this parameter is set, the regularization factors inside and
+        outside the frequency range are ignored.
+
+    Returns
+    -------
+    pyfar.Signal
+        The reslting signal after inversion.
+    """
+    if regu_final is None:
+        regu_inside = np.ones(signal.n_bins, dtype=np.double) * regu_inside
+        regu_outside = np.ones(signal.n_bins, dtype=np.double) * regu_outside
+
+        regu_final = np.ones(signal.n_bins, dtype=np.double)
+
+        idx_xfade_lower = [
+            signal.find_nearest_frequency(freq_range[0]/np.sqrt(2)),
+            signal.find_nearest_frequency(freq_range[0])]
+
+        regu_final = _cross_fade(regu_outside, regu_inside, idx_xfade_lower)
+
+        if freq_range[1] < signal.sampling_rate:
+            idx_xfade_upper = [
+                signal.find_nearest_frequency(freq_range[1]),
+                signal.find_nearest_frequency(np.min(
+                    [freq_range[1]*np.sqrt(2), signal.sampling_rate/2]))]
+
+        regu_final = _cross_fade(regu_final, regu_outside, idx_xfade_upper)
+
+        data = signal.freq
+        regu_final *= np.max(np.abs(data)**2)
+    else:
+        data = signal.freq
```

Comment:
The `else` might not be needed if moving `data = signal.freq` outside the if case

Does it have to be `signal.freq.copy()` to avoid changing the mutable np.array

Scaling with `np.max(np.abs(data)**2)` should be mentioned in the docstring. Why this is done might also be added.

Reply:
The copy is not needed here, since it's not modified.